lib/diskoverflow/sorted.go

```diff
@@ -0,0 +1,384 @@
+// Copyright (C) 2018 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package diskoverflow
+
+import (
+	"bytes"
+	"sort"
+)
+
+// Sorted stores a list of SortValue sorted by the return value of their Key method.
+// After calling PopFirst or PopLast for the first time, calls to Add will panic.
+type Sorted interface {
+	Common
+	Add(v SortValue)
+	Iter(fn func(v Value) bool, rev bool)
+	IterAndClose(fn func(v Value) bool, rev bool)
+	PopFirst() (Value, bool)
+	PopLast() (Value, bool)
+}
+
+func NewSorted(location string) Sorted {
+	s := &sorted{
+		key:      lim.register(),
+		location: location,
+	}
+	s.commonSorted = &memorySorted{key: s.key}
+	return s
+}
+
+// SortValue must be implemented by every supported type for sorting. The sorting
+// will happen according to bytes.Compare on the key.
+type SortValue interface {
+	Value
+	Key() []byte
+}
+
+type commonSorted interface {
+	common
+	add(v SortValue)
+	bytes() int64 // Total estimated size of contents
+	iter(fn func(v SortValue) bool, rev, closing bool)
+	getFirst() (SortValue, bool)
+	getLast() (SortValue, bool)
+	dropFirst() bool
+	dropLast() bool
+}
+
+type sorted struct {
+	commonSorted
+	inactive commonSorted
+	key      int
+	location string
+	spilling bool
+}
+
+func (s *sorted) Add(v SortValue) {
+	if !s.spilling && lim.add(s.key, v.Bytes()) {
+		s.inactive = s.commonSorted
+		s.commonSorted = &diskSorted{diskMap: newDiskMap(s.location)}
+	}
+	s.add(v)
+}
+
+func (s *sorted) Bytes() int64 {
+	if s.spilling {
+		return s.bytes() + lim.bytes(s.key)
+	}
+	return lim.bytes(s.key)
+}
+
+func (s *sorted) Close() {
+	s.close()
+}
+
+func (s *sorted) Iter(fn func(v Value) bool, rev bool) {
+	s.iterImpl(fn, rev, false)
+}
+
+func (s *sorted) IterAndClose(fn func(v Value) bool, rev bool) {
+	s.iterImpl(fn, rev, true)
+	s.Close()
+}
+
+func (s *sorted) iterImpl(fn func(v Value) bool, rev, closing bool) {
+	if closing {
+		defer s.Close()
+	}
+	if !s.spilling {
+		s.iter(func(v SortValue) bool {
+			return fn(v)
+		}, rev, closing)
+		return
+	}
+	aChan := make(chan SortValue)
+	iChan := make(chan SortValue)
+	go func() {
+		s.iter(func(v SortValue) bool {
+			aChan <- v
+			return true
+		}, rev, closing)
+		close(aChan)
+	}()
+	go func() {
+		s.iter(func(v SortValue) bool {
+			iChan <- v
+			return true
+		}, rev, closing)
+		close(iChan)
+	}()
+	av, aok := <-aChan
+	iv, iok := <-iChan
+	comp := -1
+	if rev {
+		comp = 1
+	}
+	for aok && iok {
+		if bytes.Compare(av.Key(), iv.Key()) == comp {
```

Comment:
This could be insertion sort I guess. Also, this has the caveat that the key has to be unique, otherwise you lose values.

Reply:
Insertion sort is just a lot slower than the current insert sequentially and sort when a value is queried for the first time. And I indeed did overlook the uniqueness problem, but that's easily solvable by adding an insertion counter at the end of the key.